v2/brokers.go

```diff
@@ -0,0 +1,232 @@
+	brokersMap := make(map[string]*broker, len(addrs))
+
+	// Ensure all connections are closed if an error occurs.
+	connectedBrokers := make([]*broker, 0, len(addrs))
+	defer func() {
+		for _, broker := range connectedBrokers {
+			broker.Close()
+		}
+	}()
+
+	for _, addr := range addrs {
+		broker, err := newBroker(ctx, addr, opts, ackReceived)
+		if err != nil {
+			return nil, err
+		}
+		brokersSlice = append(brokersSlice, broker)
+		connectedBrokers = append(connectedBrokers, broker)
+		brokersMap[addr] = broker
+	}
+
+	// All brokers are connected, disable deferred cleanup.
+	connectedBrokers = nil
+
+	return &brokers{
+		brokers:     brokersSlice,
+		brokersMap:  brokersMap,
+		ring:        hashring.New(addrs),
+		opts:        opts,
+		ackReceived: ackReceived,
+	}, nil
+}
+
+// Close closes all connections to the brokers.
+func (b *brokers) Close() error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	for _, c := range b.brokers {
+		if err := c.Close(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Update updates the connections to the brokers by closing any connection to a
+// broker that has been removed and opening connections to new brokers.
+func (b *brokers) Update(ctx context.Context, addrs []string) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// Store the new brokers in a map so we can close removed brokers.
+	newBrokersSlice := make([]*broker, 0)
+	newBrokersMap := make(map[string]*broker)
+	for _, addr := range addrs {
+		existingBroker, found := b.brokersMap[addr]
+		if found {
+			newBrokersMap[addr] = existingBroker
+			newBrokersSlice = append(newBrokersSlice, existingBroker)
+		} else {
+			broker, err := newBroker(ctx, addr, b.opts, b.ackReceived)
+			if err != nil {
+				return err
+			}
+			newBrokersMap[addr] = broker
+			newBrokersSlice = append(newBrokersSlice, broker)
+		}
+	}
+
+	for addr, broker := range b.brokersMap {
+		_, found := newBrokersMap[addr]
+		if !found {
+			// Close any removed broker.
+			broker.Close()
+		}
+	}
+
+	b.brokers = newBrokersSlice
+	b.brokersMap = newBrokersMap
+	b.ring = hashring.New(addrs)
+
+	return nil
+}
+
+// FromAddr returns an API client to a broker using its address.
+func (b *brokers) FromAddr(addr string) (proto.APIClient, error) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	broker, found := b.brokersMap[addr]
+	if !found {
+		return nil, fmt.Errorf("no broker found: %v", addr)
+	}
+
+	return broker.client, nil
+}
+
+// Random returns an API client to a random broker.
+func (b *brokers) Random() (proto.APIClient, error) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	if len(b.brokers) == 0 {
+		return nil, errors.New("no brokers")
+	}
+
+	// TODO: use a load balancer instead.
+	broker := b.brokers[rand.Intn(len(b.brokers))]
+
+	return broker.client, nil
+}
+
+// PublicationStream returns a publication stream based on a stream name and a
+// partition.
+func (b *brokers) PublicationStream(stream string, partition int32) (proto.API_PublishAsyncClient, error) {
```

Comment:
I wonder if we want to be able to say retry with any healthy broker if the correct one is down

Reply:
> I wonder if we want to be able to say retry with any healthy broker if the correct one is down

That would make sense.